src/shared/copy.c

```
@@ -795,6 +796,67 @@ static int prepare_nocow(int fdf, const char *from, int fdt, unsigned *chattr_ma
         return 0;
 }
 
+/* Copies fs-verity status.  May re-open fdt to do its job. */
+static int copy_fs_verity(int fdf, int *fdt) {
+        int r;
+
+        assert(fdf >= 0);
+        assert(fdt);
+        assert(*fdt >= 0);
+
+        r = fd_verify_regular(fdf);
+        if (r < 0)
+                return r;
+
+        struct fsverity_descriptor desc = {};
+        struct fsverity_read_metadata_arg read_arg = {
+                .metadata_type = FS_VERITY_METADATA_TYPE_DESCRIPTOR,
+                .buf_ptr = (uintptr_t) &desc,
+                .length = sizeof(desc),
+        };
+
+        r = ioctl(fdf, FS_IOC_READ_VERITY_METADATA, &read_arg);
+        if (r < 0) {
+                /* These mean that fs-verity isn't enabled on the file, filesystem, or kernel.  In all of
+                 * those cases, the correct thing to do is to do nothing at all.
+                 */
+                if (errno == ENODATA)
```

Why drop the not supported case? Don't we need to handle that as well? If we're asked to preserve fs verity status and the source filesystem doesn't support it, that needs to be a noop, not a failure.

---

I've gone back and forth on this and I came down on this side for mainly two reasons.

The first is that this code only runs in the case that it was explicitly requested with `:fsverity=copy` in the source line (or the `PRESERVE_FS_VERITY` copy flag, if someone else ever uses this).  If the source filesystem doesn't support fs-verity, but the user thinks that it ought to (since they requested the copying of this data) then there is very likely something wrong, and we'd like to hear about that rather than silently fail.

The second relates to btrfs: we use the metadata fetching ioctl, but until kernel 6.14, this wasn't supported by btrfs.  So we could have fs-verity enabled files on a btrfs filesystem (and you can even measure them) but the metadata-fetch ioctl would return `ENOTTY`.  That's a very unexpected situation and something that would definitely lead to users being very confused if we were to just fail silently.

Put another way: in light of the btrfs issue mentioned above, `ENOTTY` does *not* always mean that the filesystem lacks fs-verity support, and therefore doesn't mean that a file isn't fs-verity enabled.